### PR TITLE
Create pret-benchmark

### PR DESCRIPTION
**Pull request by Zengxianxian727**
[License: CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)

# Multi-Similarity Loss for Deep Metric Learning (MS-Loss)

This is the unofficial code with Paddlepaddle for the CVPR 2019 paper [Multi-Similarity Loss with General Pair Weighting for Deep Metric Learning](http://openaccess.thecvf.com/content_CVPR_2019/papers/Wang_Multi-Similarity_Loss_With_General_Pair_Weighting_for_Deep_Metric_Learning_CVPR_2019_paper.pdf)

And the official code is [here](https://github.com/MalongTech/research-ms-loss)

## Initial Paddlepaddle 1.8.x
pip install paddlepaddle-gpu

### Prepare the data and the pretrained model 

The following script will prepare the [CUB](http://www.vision.caltech.edu.s3-us-west-2.amazonaws.com/visipedia-data/CUB-200-2011/CUB_200_2011.tgz) dataset for training by downloading to the ./resource/datasets/ folder; which will then build the data list (train.txt test.txt):

```bash
./scripts/prepare_cub.sh
```

Download the imagenet pretrained model of 
[bninception](http://data.lip6.fr/cadene/pretrainedmodels/bn_inception-52deb4733.pth) and put it in the folder. And then we should use it to generate the Paddlepaddle pretrained model.
```bash
python ./pretrained_model/pytorch2paddle.py
```

### Installation

```bash
python setup.py develop build
```
###  Train and Test on CUB200-2011 with MS-Loss

```bash
./scripts/run_cub.sh
```
Trained models will be saved in the ./output/ folder if using the default config.

Best recall@1 is 65.1 (65.7 in the paper).

### Contact

For any questions, please feel free to reach 
```
zengxianxian727@foxmail.com
```

### Citation

If you use this method or this code in your research, please cite as:

    @inproceedings{wang2019multi,
    title={Multi-Similarity Loss with General Pair Weighting for Deep Metric Learning},
    author={Wang, Xun and Han, Xintong and Huang, Weilin and Dong, Dengke and Scott, Matthew R},
    booktitle={Proceedings of the IEEE Conference on Computer Vision and Pattern Recognition},
    pages={5022--5030},
    year={2019}
    }

## License

MS-Loss is CC-BY-NC 4.0 licensed, as found in the [LICENSE](LICENSE) file. It is released for academic research / non-commercial use only. If you wish to use for commercial purposes, please contact bd@malong.com.

### CODE REVIEW

**Comment by Zengxianxian727** [comment]
[research-ms-loss-paddle.zip](https://github.com/PaddlePaddle/Contrib/files/5181128/research-ms-loss-paddle.zip)
